luqum/utils.py

```diff
@@ -80,6 +80,128 @@ def __call__(self, tree):
+            yield from self.generic_visit(node, context)
+            return
+
+        new_node = AndOperation(pos=node.pos, size=node.size, head=node.head, tail=node.tail)
+        new_node.children = ()
+
+        # We collect all ranges from a AND's children with the same bound side, and pop the
+        # first one from the list when we encounter a different bound side. This allows us
+        # to join [x TO *] AND [y TO *] to [* TO a] AND [* TO b] as [x TO a] AND [y TO b]
+        # We adjust the first enounter of the range, and do not add the joined one.
+
+        # Note: possible_ranges_bound_side has no meaning if possible_ranges is empty
+        possible_ranges, possible_ranges_bound_side = [], None
+        for child in self.clone_children(node, new_node, context):
+            # Determines the bound side of the range (high or low), meaning that the other side is
+            # a wildcard. Assigns None if the provided node is not a Range, bound on both sides,
+            # or bound on neither side.
+            child_bound_side = None
+            if isinstance(child, Range):
+                if child.low == self.WILDCARD_WORD and child.high != self.WILDCARD_WORD:
+                    child_bound_side = 'high'
+                elif child.low != self.WILDCARD_WORD and child.high == self.WILDCARD_WORD:
+                    child_bound_side = 'low'
+
+            if child_bound_side is not None:
+                # If we have encountered a Range with only one bound side, we may be able to join it
+                if not possible_ranges or possible_ranges_bound_side == child_bound_side:
+                    # We have not yet encountered any unbound ranges, or the encountered unbound
+                    # rangesare all the same direction, so store a pointer to this child and we may
+                    # be able to join a future range. The flow will continue.
+                    possible_ranges.append(child)
+                    # This assignment only has meaning if possible_ranges was empty
+                    possible_ranges_bound_side = child_bound_side
+                else:
+                    # There is at least one possible range to join to. We adjust the first range
+                    # encountered of the different side
+                    joining_child = possible_ranges.pop(0)
+                    if child_bound_side == 'low':
+                        joining_child.low = child.low
+                        joining_child.include_low = child.include_low
+                    else:
+                        joining_child.high = child.high
+                        joining_child.include_high = child.include_high
+                    # we should not need to adjust the head/tail because those would be included
+                    # from the child range we are joining from
+
+                    # do not add the child we have just merged from
+                    continue
+
+            new_node.children += (child, )
```

Comment:
I think that the original code is better, since the list will only contain one side left unbounded. I have added more comments to make this more clear. 

If you feel that we should still rewrite, we could do that, but I'm not sure this is the clearer option.